apps/src/codebridge/FileBrowser/dragAndDropUtils.ts

```diff
@@ -0,0 +1,188 @@
+import {DEFAULT_FOLDER_ID} from '@codebridge/constants';
+import {
+  KeyboardCode,
+  KeyboardCoordinateGetter,
+  CollisionDetection,
+  rectIntersection,
+  Collision,
+  Active,
+  ClientRect,
+} from '@dnd-kit/core';
+import {RectMap} from '@dnd-kit/core/dist/store';
+
+import {ProjectFolder} from '@cdo/apps/lab2/types';
+
+import {getFolderChildren} from '../utils/getFolderChildren';
+
+import {DragType} from './types';
+
+export const FOLDER_DROP_OFFSET = 16;
+
+// Custom keyboard coordinate getter for the file browser.
+// When we are moving an item via the keyboard, we move it to the next available folder
+// up/down in the browser, rather than just a flat number of pixels.
+// If we are moving a folder, we skip any folders that are children of the current folder.
+export const fileBrowserKeyboardCoordinateGetter: (
+  folders: Record<string, ProjectFolder>
+) => KeyboardCoordinateGetter = folders => (event, args) => {
+  if (event.code !== KeyboardCode.Up && event.code !== KeyboardCode.Down) {
+    return;
+  }
+  event.preventDefault();
+
+  const {context} = args;
+  const {droppableRects, over, active} = context;
+  let orderedRects = Array.from(droppableRects.keys());
+  const currentId = active?.data.current?.id;
+  const dragType = active?.data.current?.type;
+
+  if (dragType === DragType.FOLDER) {
+    // We can't move a folder into a folder that is a child of itself
+    const allChildren = getFolderChildren(
+      currentId as string,
+      Object.values(folders)
+    );
+    orderedRects = orderedRects.filter(
+      id => !allChildren.includes(id as string)
+    );
+  }
+
+  // Sort the available droppable rectangles by their top coordinate, except
+  // for the root folder (DEFAULT_FOLDER_ID), which should always be last, as it
+  // is the folder that contains all the other folders.
+  orderedRects.sort((a, b) => {
+    // Default folder is always last in the list.
```

Comment:
I'm curious about why the default folder is last in the list, but perhaps it doesn't mean 'visually last'

Reply:
Default folder is always "visually last" because it's the containing folder of the entire file browser. It's an implicit folder, the users don't see it. In this screenshot everything is in the default folder.
![Screenshot 2025-04-22 at 3 03 14 PM](https://github.com/user-attachments/assets/26d28ad3-eca5-4a11-a489-b1217d3d59fa)